### ISSUE TITLE
Can papers be associated with individual Pull Requests?

### ISSUE THREAD

**Issue by jakevdp** [openjournals/joss#52]
This was an idea brought up by @arokem here at UW. He mentioned the common issue of assigning credit to contributions within individual projects...

For example, say I commit a new feature/algorithm to the astropy project, or even do a major refactoring of part of the package. Can we offer a way to get a citeable DOI for that work?

**Comment by arokem** [comment]
Thanks @jakevdp for raising this. This is a (potential) authors request! :-)

I am considering whether we could publish things like this PR: https://github.com/nipy/dipy/pull/740, on a project for which there is already a software paper out (http://journal.frontiersin.org/article/10.3389/fninf.2014.00008/full), but the author of this PR is not an author on that publication.

But it's completely understandable if this discussion lingers until after you've launched and gotten this thing going (congratulations!).

**Comment by arokem** [comment]
I'll just add that one consideration is credit assigned to code reviewers within these kinds of projects. This is a selfish consideration: I expend a significant amount of my OSS energy reviewing code. 

My understanding is that anyone can be an author on a JOSS paper, even if there are no commits under their user-name in the relevant repo. Is  that correct?  

Another (semi-similar) use-case are PIs who "provide guidance" (=write grants to fund the work ;-D).

**Comment by crawfordsm** [comment]
As the maintainer on the astropy.stats package, I would love if major contributions could get their own submission and be scientifically peer reviewed.   Code review is being carried out pretty well within the project, but having experts in the field comment on how the code is performing or is described is lacking and not straight forward to do within the project.  So this would be a real service to me as a maintainer!

Plus, it would provide an obvious way for people making significant contributions to a large project to be properly credited for their work.  People could list on their CV they were reviewers for #JOSS which at least counts for something (although I agree reviewers are not credited enough).

**Comment by tracykteal** [comment]
In the astropy case, would examples of major contributions be a new feature or the better/faster/smarter implementation of an existing algorithm? In those cases, I could see how those were standalone submissions. That new feature or algorithm implementation would have its own documentation, tests, example usage, statement of need, functionality and performance. It does look like https://github.com/nipy/dipy/pull/740 could meet this criteria. 

@arfon @danielskatz what about having them submit it and see if it works with the existing review process, with the understanding that if it doesn't, they might have to wait a bit until we can revisit?

I think authorship lists can be decided by the submitters whether that's based on PRs, intellectual contribution, code review or other factors the project considers important.

**Comment by jakevdp** [comment]
@tracykteal – Good thoughts. I think one challenge is getting a DOI for such a contribution. The github/zenodo process wouldn't really apply, unless we figure out a way to tweak it.

**Comment by tracykteal** [comment]
Right, since they're linked to a release. I agree we can't do that tweak right now, so it might be worth holding off on implementing this (as had already been suggested).

**Comment by danielskatz** [comment]
My thoughts:

The overall model for recognizing contributions seems to be that once we have enough new contributions to create a new release, we do so, and the new release can get a new DOI, and could be the subject of a new JOSS paper.

I agree with @tracykteal that authorship is determined by the submitter.  I think the submitter should include as authors all who have contributed to that release, which is probably all people who contributed to the previous releases (maybe unless their additions have completely been replaced, but even then, they made a contribution to the project, even without having written any of the existing code) plus the people who have made contributions since the previous release.

Maybe I'm missing something, but I don't see that there's a problem that we can't solve using the existing mechanisms.

**Comment by arokem** [comment]
On Wed, May 11, 2016 at 5:25 PM, Daniel S. Katz <notifications@github.com>
wrote:

> My thoughts:
>
> The overall model for recognizing contributions seems to be that once we
> have enough new contributions to create a new release, we do so, and the
> new release can get a new DOI, and could be the subject of a new JOSS paper.
>
> I agree with @tracykteal <https://github.com/tracykteal> that authorship
> is determined by the submitter. I think the submitter should include as
> authors all who have contributed to that release, which is probably all
> people who contributed to the previous releases (maybe unless their
> additions have completely been replaced, but even then, they made a
> contribution to the project, even without having written any of the
> existing code) plus the people who have made contributions since the
> previous release.
>
> Maybe I'm missing something, but I don't see that there's a problem that
> we can't solve using the existing mechanisms.
>
> Releases made up of multiple complex, but distinct PRs (or series of PRs)
on different topics:

For example, this is one distinct feature mostly led by one contributor:

- https://github.com/nipy/dipy/pull/740

And these two are a collection of features mostly led by another
contributor:

- https://github.com/nipy/dipy/pull/828
- https://github.com/nipy/dipy/pull/762

These will presumably be released together, but I think they should go into
two different papers. Otherwise, we might be incentivizing splitting up big
projects into multiple tiny ones, just for the DOI, effectively making the
software maintenance more difficult.

**Comment by jakevdp** [comment]
> Otherwise, we might be incentivizing splitting up big projects into multiple tiny ones, just for the DOI, effectively making the software maintenance more difficult.

This has actually crossed my mind... in some sense by contributing my current Lomb-Scargle PR to astropy, I'm going to greatly reduce the utility of my ``astroml`` and ``gatspy`` packages, which will reduce my citations on their DOIs & associated papers. I've decided for the moment it's more important to serve the astropy community well than to bump my citation rates, but that's only a rational decision for me because I have some freedom in my current position to eschew the academic reward structure as a matter of principle :smile: I wouldn't expect that of others on a more typical academic career path.

**Comment by arokem** [comment]
OK - we now have a concrete case for this

This is the branch on which the new model was developed: https://github.com/nipy/dipy/pull/1110
This is the branch with the paper: https://github.com/nipy/dipy/pull/1126

How should we go about submitting this as a JOSS paper?

**Comment by arfon** [comment]
> OK - we now have a concrete case for this

👍  excellent! 

> How should we go about submitting this as a JOSS paper?

To be honest I'm not exactly sure 😕 . @arokem - nipy/dipy#1110 looks to already be in `master` and presumably the paper in nipy/dipy#1126 will end up end up in a paper _specific to this contribution_ rather than a general paper about https://github.com/nipy/dipy?

Am I correct in assuming that there's no archive DOI (e.g. Zenodo/figshare) for the top level https://github.com/nipy/dipy project? The reason I ask is that in https://github.com/codemeta/codemeta/issues/91 there is a good discussion of the DataCite relationships and I think any DOI/paper associated with nipy/dipy#1110 `IsPartOf` the https://github.com/nipy/dipy project and we need to find a way to capture this relationship in the JOSS submission.

I would love to get a sanity check on this from @mfenner as to whether this sounds like a reasonable approach? i.e. Archive with Zenodo and assign a DOI for https://github.com/nipy/dipy and then create a new DOI for nipy/dipy#1110 and add `IsPartOf` to the metadata.

**Comment by arokem** [comment]
Thanks Arfon!

On Tue, Oct 25, 2016 at 6:35 PM, Arfon Smith <notifications@github.com>
wrote:

> OK - we now have a concrete case for this
>
> 👍 excellent!
>
> How should we go about submitting this as a JOSS paper?
>
> To be honest I'm not exactly sure 😕 . @arokem <https://github.com/arokem>
> - nipy/dipy#1110 <https://github.com/nipy/dipy/pull/1110> looks to
> already be in master and presumably the paper in nipy/dipy#1126
> <https://github.com/nipy/dipy/pull/1126> will end up end up in a paper *specific
> to this contribution* rather than a general paper about
> https://github.com/nipy/dipy?
>
> That's correct. We are interested in a paper specifically about this
contribution (A Google Summer of Code project, incidentally).

> Am I correct in assuming that there's no archive DOI (e.g.
> Zenodo/figshare) for the top level https://github.com/nipy/dipy project?
> The reason I ask is that in codemeta/codemeta#91
> <https://github.com/codemeta/codemeta/issues/91> there is a good
> discussion of the DataCite relationships and I think any DOI/paper
> associated with nipy/dipy#1110 <https://github.com/nipy/dipy/pull/1110>
> IsPartOf the https://github.com/nipy/dipy project and we need to find a
> way to capture this relationship in the JOSS submission.
>
There is no Zenodo entry, but we should make one. Would we need to make a
new release for the Zenodo DOI to get generated? That is my understanding,
but I wanted to make sure. It's high time we made a release anyway, but I
wanted to make sure.

> I would love to get a sanity check on this from @mfenner
> <https://github.com/mfenner> as to whether this sounds like a reasonable
> approach? i.e. Archive with Zenodo and assign a DOI for
> https://github.com/nipy/dipy and then create a new DOI for nipy/dipy#1110
> <https://github.com/nipy/dipy/pull/1110> and add IsPartOf to the metadata.
>
I've seen this IsPartOf being implemented very nicely here:
http://gigadb.org/dataset/100215, and that seems like exactly what we would
like for this as well. All papers related to a particular software project
can then be related to each other through the project DOI. I see a lot of
benefits to that, if its possible to implement (i.e. if the metadata is
editable by me to add the `IsPartOf` field).

Thanks!

**Comment by sahmed95** [comment]
Gentle reminder. 

The reward and joy of sharing your code with others and to see it being used to fuel further research is really something than can't be quantified but citations seem to be the currency needed for advancing in academia.

Something like this would really help a lot of people and perhaps encourage more contributions to open source scientific computing.

**Comment by kyleniemeyer** [comment]
@sahmed95 the model that we have settled on for now is what @danielskatz describes above:
> The overall model for recognizing contributions seems to be that once we have enough new contributions to create a new release, we do so, and the new release can get a new DOI, and could be the subject of a new JOSS paper.

Meaning, new contributions to a software package can certainly be recognized, but through the submission of the whole software + new article for review, for a new major version, for example. This allows new contributors to be recognized for their efforts.

**Comment by sahmed95** [comment]
Thanks for the reply. This is great. So, between every few major releases, the community can decide if they want a new JOSS paper and include all the merged code and acknowledge the contributers. So something like v2.0 can have a doi and say after some time when v3.2 releases, another JOSS article can be submitted which adds contributions of the new people since v2.0 and gets another doi?

**Comment by kyleniemeyer** [comment]
In theory, yes. The author list could easily be extended for a new article.

For a new JOSS article, the new version should represent a major release compared with the version described in the prior article, whether due to a substantial change in features, capabilities, performance, etc. Whether the new version is suitable (meaning, is there enough new content represented in the software) would be up to the reviewers and editor of a submission.

**Comment by sahmed95** [comment]
But the way JOSS papers are published just gives information about the main repository and what it does. When new modules get added, how would those be highligted ? Because some contributers will probably just have simple bug fixes while others may be adding complete new modules. Will it be something like the changelogs listing down the major additions and only acknowledging those contributers ?

**Comment by kyleniemeyer** [comment]
Yes, the JOSS papers describe the overall software, but a new paper would likely need to describe the changes from a prior version (just like any research paper building on past work). This should be supplemented with a detailed CHANGELOG in the software repo itself. 

There currently isn't any requirement (or mechanism) to highlight the contributions of individual authors, so adding their names to the list of authors—which we leave up to the project team—would be the appropriate way to acknowledge new contributors. The new paper would probably need to include all the old authors and the new authors, since the work of all is being represented by the paper.

**Comment by sahmed95** [comment]
Great. Thanks. Having a link to the changelogs could help but I guess it's okay just to have the author names added. The description would anyway mention the big changes. But another issue might be that the same authors will now be mentioned in multiple JOSS papers even though they might not have contributed anything new to the subsequent versions. Is that an issue?

**Comment by kyleniemeyer** [comment]
No, I don't see this as an issue. People using the new version should cite the newer paper, and the original efforts of the "older" contributors are still important and should be credited.

**Comment by sahmed95** [comment]
Great. Thank you very much.

**Comment by labarba** [comment]
If the original authors made no contributions to the major new release, they don't need to be added as authors to the second paper. This is up to the project to figure out.

**Comment by danielskatz** [comment]
I'm going to close this, as I don't think we're going to change the JOSS model at this point, which seems to handle most of the common use cases for new versions and contributions. It can be re-opened if needed.
